src/autopas/selectors/tuningStrategy/PredictiveTuning.h

```diff
@@ -0,0 +1,370 @@
+                                           const std::set<Newton3Option> &allowedNewton3Options) {
+  // generate all potential configs
+  for (auto &containerOption : allowedContainerOptions) {
+    // get all traversals of the container and restrict them to the allowed ones
+    const std::set<TraversalOption> &allContainerTraversals =
+        compatibleTraversals::allCompatibleTraversals(containerOption);
+    std::set<TraversalOption> allowedAndApplicable;
+    std::set_intersection(allowedTraversalOptions.begin(), allowedTraversalOptions.end(),
+                          allContainerTraversals.begin(), allContainerTraversals.end(),
+                          std::inserter(allowedAndApplicable, allowedAndApplicable.begin()));
+
+    for (auto &cellSizeFactor : allowedCellSizeFactors)
+      for (auto &traversalOption : allowedAndApplicable) {
+        for (auto &dataLayoutOption : allowedDataLayoutOptions) {
+          for (auto &newton3Option : allowedNewton3Options) {
+            _searchSpace.emplace(containerOption, cellSizeFactor, traversalOption, dataLayoutOption, newton3Option);
+          }
+        }
+      }
+  }
+
+  AutoPasLog(debug, "Points in search space: {}", _searchSpace.size());
+
+  if (_searchSpace.empty()) {
+    autopas::utils::ExceptionHandler::exception("PredictiveTuning: No valid configurations could be created.");
+  }
+
+  for (auto &configuration : _searchSpace) {
+    std::vector<std::pair<int, size_t>> vector;
+    _traversalTimesStorage.emplace(configuration, vector);
+  }
+
+  _currentConfig = _searchSpace.begin();
+}
+
+void PredictiveTuning::selectPredictedConfigurations() {
+  if (_searchSpace.size() == 1 || _timer == 0 || _timer == 1) {
+    _optimalSearchSpace = _searchSpace;
+    return;
+  }
+
+  predictConfigurations();
+
+  auto optimum = std::min_element(_traversalPredictions.begin(), _traversalPredictions.end(),
+                                  [](std::pair<Configuration, size_t> a, std::pair<Configuration, size_t> b) -> bool {
+                                    return a.second < b.second;
+                                  });
+
+  _optimalSearchSpace.emplace(optimum->first);
+
+  // selects configurations that are near the optimal prediction or have not been tested in the last five iterations
+  for (auto &configuration : _searchSpace) {
+    auto vector = _traversalTimesStorage[configuration];
+    if ((_timer - vector[vector.size() - 1].first) >= 5 ||
+        ((float)_traversalPredictions[configuration] / optimum->second <= 1.2)) {
+      _optimalSearchSpace.emplace(configuration);
+    }
+  }
+}
+
+void PredictiveTuning::predictConfigurations() {
+  linePrediction();
+  /*switch (method) {
+  * case line: linePrediction(); break;
+  * default:
+
+  }*/
+}
+
+void PredictiveTuning::linePrediction() {
+  for (auto &configuration : _searchSpace) {
+    auto vector = _traversalTimesStorage[configuration];
```

Comment:
```suggestion
    const auto &vector = _traversalTimesStorage[configuration];
```
also `const` everything else that can be `const`